### PR TITLE
Event emitter improvements

### LINKED ISSUE #96: EventEmitter Memory Leak
### Description

```sh
(node:93636) MaxListenersExceededWarning: Possible EventEmitter memory leak detected.
11 disconnect listeners added to [Socket]. Use emitter.setMaxListeners() to increase limit
```

### Epic/Story

_No response_

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### LINKED ISSUE #102: widget-action and other events are increasing with each deploy
Due to the organisation of code, events like `widget-action` cannot be unhooked and as a result, each time a node is deployed, new events are hooked up causing duplicate, triplicate messages being posted to node-red

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #96 
closes #102 

## Description

* uses common handlers for the 3 widget socket.io channels - instead of - an event hander per widget per handler
* ensures clean up socket io handlers on disconnect / close
* ensures emit config is requested when a node is added or removed in all deploy modes
* updates docs
* improved jsdoc type hints in ui-base
* removes the need for tracking load/change/action events
* only emits config to "own nodes connections"
   * prior, all connections (regardless of which base they belong to) were issued a config when any 1 base was updated. Now, only sockets connected to `this` base are emitted new config

## Related Issue(s)

#96 
#102 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: nodes/config/ui_base.js [hunk 1]
```diff
@@ -327,144 +517,37 @@ module.exports = function (RED) {
...
+                changes = true
+            }
+            if (changes) {
+                node.requestEmitConfig()
             }
         }
+
+        // Finally, queue up a config emit to the UI.
+        // NOTE: this is a cautionary measure only - typically the registration of nodes will queue up a config emit
+        // but in the event no widgets are registered, we still need still emit a config
```
Comment (Steve-Mcl) [inline nodes/config/ui_base.js:549]:
```suggestion
        // but in cases where the dashboard has no widgets registered, we still need to emit a config
```

**Comment by joepavitt** [comment]
In testing found a couple of issues. If you wire a text input into a text node:

Screenshot 2023-08-05 at 10 09 12

Then the full `msg` object is displayed, not just the payload as should be the default:

Screenshot 2023-08-05 at 10 09 25

This is odd however, because the Text widget does specify:

```vue
value: function () {
    return this.messages[this.id]?.payload
}
```

If we then refresh the view, the value in the text input, changes to an object, not the value input in the first place:

Screenshot 2023-08-05 at 10 10 32

**Comment by joepavitt** [comment]
Seems like a dodgy nesting issue, this is the output if we add:

```js
console.log(this.messages[this.id])
```

to the computed `value` of the Text widget.

Screenshot 2023-08-05 at 10 13 06

**Comment by joepavitt** [comment]
> Seems like a dodgy nesting issue

This should be fixed in https://github.com/flowforge/flowforge-nr-dashboard/pull/105/commits/89afb65f9a0f7779b39f165ec98e125df596a09a

**Review by joepavitt — APPROVED** [review]

**Comment by Steve-Mcl** [comment]
Thanks for the catch Joe.

**Pull request merged**